src/sparse/impl/KokkosSparse_twostage_gauss_seidel_impl.hpp

```diff
@@ -772,134 +899,262 @@ namespace KokkosSparse{
           KokkosKernels::Impl::zero_vector<x_value_array_type, execution_space>(nrhs, localX);
         }
         for (int sweep = 0; sweep < NumSweeps; ++sweep) {
-          // R = B - A*x
+          bool forward_sweep = (direction == GS_FORWARD ||
+                               (direction == GS_SYMMETRIC && sweep%2 == 0));
+          // compute residual vector
           KokkosBlas::scal (localR, one, localB);
           if (sweep > 0 || !init_zero_x_vector) {
-            KokkosSparse::
-            spmv ("N", scalar_t(-one), crsmatA,
-                                       localX,
-                                 one,  localR);
+            if (compact_form) {
+              if (forward_sweep) {
+                // R = B - U*x
+                KokkosSparse::
+                spmv ("N", scalar_t(-one), crsmatUa,
+                                           localX,
+                                     one,  localR);
+              } else {
+                // R = B - L*x
+                KokkosSparse::
+                spmv ("N", scalar_t(-one), crsmatLa,
+                                           localX,
+                                     one,  localR);
+              }
+              if (omega != one) {
+                // R = B - (U + (1-1/omega)D)*x
+                scalar_t omega2 = (one/omega - one);
+                auto localY = Kokkos::subview (localX, range_type(0, num_rows), Kokkos::ALL ());
+                KokkosBlas::mult (zero, localZ,
+                                  one,  localDa, localY);
+                KokkosBlas::axpy (omega2, localZ, localR);
+              }
+            } else { // not compact_form
+              // R = B - A*x
+              KokkosSparse::
+              spmv ("N", scalar_t(-one), crsmatA,
+                                         localX,
+                                   one,  localR);
+#ifdef KOKKOSSPARSE_IMPL_TIME_TWOSTAGE_GS
+              {
+                auto localRj = Kokkos::subview (localR, Kokkos::ALL (), range_type (0, 1));
+                single_vector_view_t Rj (localRj.data (), num_rows);
+                std::cout << "norm(GS)-" << sweep << " " << KokkosBlas::nrm2 (Rj)
+                          << " (" << (forward_sweep ? "forward" : "backward" )  << ")"
+                          << std::endl;
+              }
+#endif
+            }
           }
-          if (!two_stage) { // ===== sparse-triangular solve =====
-            if (direction == GS_FORWARD ||
-               (direction == GS_SYMMETRIC && sweep%2 == 0)) {
-              // Z = (L+D)^{-1} * R
+          //if (sweep == 0) {
+          //  auto localY = Kokkos::subview (localX, range_type(num_rows, localX.extent(0)), Kokkos::ALL ());
+          //  Kokkos::deep_copy (localY, zero);
+          //}
+          if (!two_stage) {
+            // ===== sparse-triangular solve =====
+            // TODO: omega is not supported here (because omega*L + D is extracted in initialize_numeric, but omega is passed into apply)
```

Comment:
Again, I think it's fine to deal with ``omega != 1.0`` during numeric rather than apply. Because of that, it would make sense to go through the comments here like ``Z = (omega * L + D)^{-1} * R`` and just not mention omega, since ``crsmatU`` is already ``omega * L + D``. Alternately, you could clarify by saying ``Z = (omega * L + D)^{-1} * R = crsmatU^{-1} * R``.